### PR TITLE
Add regression test for SafeSocketHandle getting collected

### PR DESCRIPTION
**Pull request by stephentoub**
Contributes to https://github.com/dotnet/corefx/issues/37044
cc: @wfurt

### CODE REVIEW

File: src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Xunit;
+
+namespace System.Net.Sockets.Tests
+{
+    public partial class DisposedSocket
+    {
+        //[ActiveIssue(37044, TestPlatforms.AnyUnix)]
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs]: I will uncomment this once I verify via CI that it does indeed fail on all of our tested Unix platforms.

**Comment by wfurt** [comment]
It passed on my ubuntu 18 system
```
System.Net.Sockets.Tests.DisposedSocket.NonDisposedSocket_SafeHandlesCollected [FINISHED] Time: 0.0146968s
```

File: src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                GC.Collect();
+                Assert.Equal(0, handles.Count(h => h.IsAlive));
+            });
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static List<WeakReference> CreateHandles()
+        {
+            var handles = new List<WeakReference>();
```
Comment (tmds) [inline src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs]: @stephentoub don't we need something 'async' to get into the static SocketAsyncEngine Dictionary?

Comment (stephentoub) [reply src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs]: Yup, forgot that part.  Thanks.

**Comment by stephentoub** [comment]
This is now failing as expected on Unix:
```
Unhandled Exception of Type System.AggregateException
Message :
System.AggregateException : One or more errors occurred. (Assert.Equal() Failure
Expected: 0
Actual:   100) (Assert.Equal() Failure
Expected: 0
Actual:   100) (Assert.Equal() Failure
Expected: 0
Actual:   100) (Assert.Equal() Failure
Expected: 0
Actual:   100) (Assert.Equal() Failure
Expected: 0
Actual:   100)
---- Assert.Equal() Failure
Expected: 0
Actual:   100
---- Assert.Equal() Failure
Expected: 0
Actual:   100
---- Assert.Equal() Failure
Expected: 0
Actual:   100
---- Assert.Equal() Failure
Expected: 0
Actual:   100
---- Assert.Equal() Failure
Expected: 0
Actual:   100
Stack Trace :
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 45
   at System.Net.Sockets.Tests.DisposedSocket.NonDisposedSocket_SafeHandlesCollected(Boolean clientAsync) in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 22
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace #1 (Xunit.Sdk.EqualException) -----
   at System.Net.Sockets.Tests.DisposedSocket.<>c__DisplayClass112_0.<NonDisposedSocket_SafeHandlesCollected>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 26
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 37
----- Inner Stack Trace #2 (Xunit.Sdk.EqualException) -----
   at System.Net.Sockets.Tests.DisposedSocket.<>c__DisplayClass112_0.<NonDisposedSocket_SafeHandlesCollected>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 26
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 37
----- Inner Stack Trace #3 (Xunit.Sdk.EqualException) -----
   at System.Net.Sockets.Tests.DisposedSocket.<>c__DisplayClass112_0.<NonDisposedSocket_SafeHandlesCollected>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 26
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 37
----- Inner Stack Trace #4 (Xunit.Sdk.EqualException) -----
   at System.Net.Sockets.Tests.DisposedSocket.<>c__DisplayClass112_0.<NonDisposedSocket_SafeHandlesCollected>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 26
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 37
----- Inner Stack Trace #5 (Xunit.Sdk.EqualException) -----
   at System.Net.Sockets.Tests.DisposedSocket.<>c__DisplayClass112_0.<NonDisposedSocket_SafeHandlesCollected>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs:line 26
   at System.RetryHelper.Execute(Action test, Int32 maxAttempts, Func`2 backoffFunc) in /_/src/CoreFx.Private.TestUtilities/src/System/RetryHelper.cs:line 37
```
I'll uncomment the ActiveIssue attribute so it can be merged.

**Pull request merged**